[#63550] Allow editing of individual work package/project attributes

dombesz opened the pull request:
# Ticket
https://community.openproject.org/work_packages/63550

# What are you trying to accomplish?
Solve the issue where customizable objects cannot be saved if there is an invalid custom field associated with them.
When custom fields are displayed in the form being saved, this behaviour is normal. When custom fields are not displayed on the saving form, validating them is causing unexpected validation errors, and the user cannot know why the save has failed.

# What approach did you choose and why?

Validate only the custom fields that are being displayed on the saving form.

Todo list:
- [x] Implement the new custom field validation logic.
- Update model specs for:
  - [x] Project
  - [x] WorkPackage
  - [x] Users
  - [x] Groups
  - [x] Versions
  - [x] TimeEntries
- Update Feature specs for:
  - [x] Projects
  - [x] WorkPackages
  - [x] Users
  - [x] Groups
  - [x] Versions
  - [x] TimeEntries
- Update Api specs for:
  - [x] Projects
  - [x] WorkPackages
  - [x] Users
  - [x] Groups
  - [x] Versions
  - [x] TimeEntries
  - [X] Additionally solves ["Required for all projects" should not require a value to be set for a Calculated Value](https://community.openproject.org/wp/67839)

- Out of scope:
  - [X] Remove section based validation on the project attributes form.
  - [ ] Remove the `:saving_custom_fields` context.

# Merge checklist

- [x] Added/updated tests
- [x] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

dombesz committed:
Add custom field testcases to user create/update form/resource specs

dombesz committed:
[#63550] Allow editing of individual work package/project attributes even if certain other attributes are invalid (eg. required field empty)

https://community.openproject.org/work_packages/63550

dombesz committed:
Remove unused code

dombesz committed:
Wp and group aac spec works

dombesz committed:
User cf spec works

dombesz committed:
Version cf specs works

dombesz committed:
Project cf spec works

dombesz committed:
Add project overview specs

dombesz committed:
Create group custom field validation specs

dombesz committed:
Implement custom field via the groups api, custom field api validations specs.

dombesz committed:
Add version resource spec

dombesz committed:
Add version create update including custom fields integration specs.

dombesz committed:
Complete group resource specs.

dombesz committed:
Update the SetAttributesService to handle different custom field setters when finding cfs to validate.

dombesz committed:
Add notes to remove temporary fixes.

dombesz committed:
Use custom_value_for accessors instead of iterating custom field values in specs

dombesz committed:
Add work package api resource specs

dombesz committed:
Add project custom fields api resource specs

dombesz committed:
Make new work package custom field validation feature spec a bit more descriptive.

dombesz committed:
Add time entry model acts as customizable specs.

Add time entry model custom field specs

dombesz committed:
Add time entry create api custom field specs

dombesz committed:
Add time entry dialog specs

dombesz committed:
Add User invite and ldap registration specs required custom field specs

dombesz committed:
Add user invite modal specs for missing required custom fields while for both users and groups invited.

dombesz committed:
Add project form resource specs

dombesz committed:
Adjust structure on work package create and update resource specs

dombesz committed:
Add work package create and update form resource specs

dombesz committed:
Update project create, update, form resources specs

dombesz committed:
Add version form resource specs

dombesz committed:
Refactor group_resource_spec custom field specs

dombesz committed:
Update all project resource specs to have custom field specs

dombesz committed:
Update custom field specs time entries resource specs.

dombesz committed:
Add admin only and required custom field specs to the project resource specs

dombesz committed:
Fix group representer and versions update form resource specs.

dombesz committed:
Fix specs by triggering custom field validation.

dombesz committed:
Fix aac expectations

dombesz committed:
Fix more failing specs

dombesz committed:
Update customizable specs

dombesz committed:
Fix delete service spec

dombesz committed:
Do not clear the custom_values_to_validate in the set attribute service, if there are no custom fields to be validated.

This behaviour change allows the custom_values_to_validate array to be
set in other services that are using the WorkPackage::UpdateService.
Currently only the CustomActions::UpdateWorkPackageService is using the
validation array.

dombesz committed:
Try fixing spec.

dombesz committed:
Fix specs, activate all cusotm fields validations on the work packages update api endpoint.

dombesz committed:
Introduce EditFormWorkPackageChangeset to enforce custom field validation when editing work packages in angular.

dombesz committed:
Remove temporary fix.

dombesz committed:
Fix typo

dombesz committed:
Validate the register user service

dombesz committed:
Add specs with forced password change having invalid user custom fields

dombesz committed:
Add validate custom fields flag for the work packages api to allow explicit validation too

dombesz committed:
Make the custom field validation flag configurable on the resource changeset.

dombesz committed:
[#67839] "Required for all projects" should not require a value to be set for a Calculated Value

https://community.openproject.org/work_packages/67839

dombesz committed:
Fix ldap signing in spec

dombesz committed:
Do not validate custom fields when executing a custom action.

dombesz committed:
Validate changed custom fields when executing custom actions

dombesz committed:
Validate custom fields in all the custom actions that are custom field related.

dombesz committed:
Do not use a set, just make the custom_values_to_validate uniq.

dombesz committed:
Fix rubocop complaint.

dombesz committed:
Add api documentation

dombesz committed:
Fix rubocop issues.

dombesz committed:
Fix eslint issue

ulferts commented on app/models/custom_actions/actions/strategies/custom_field.rb:
```diff
@@ -30,7 +30,10 @@
+    if work_package.respond_to?(custom_field.attribute_setter)
+      set_custom_field_value(work_package)
+      validate_custom_field(work_package)
+    end
```
These lines are repeated a couple of times in the application. Maybe a chance for extracting them into a module.

dombesz replied:
Will be addressed separately.

ulferts commented on app/services/work_packages/set_attributes_service.rb:
```diff
@@ -54,6 +57,16 @@ def mark_templated_subject
+    else
+      super(attributes)
+    end
+  end
```
Is this the only model one can pass the `validate_custom_fields` param to?

If not, the method could be moved to `super`

dombesz replied:
Currently, only the work packages support the `validate_custom_fields` param. Maybe we can extend it in the future if we need to.

ulferts commented on lib_static/plugins/acts_as_customizable/lib/acts_as_customizable.rb:
```diff
@@ -220,35 +220,30 @@ def reset_custom_values!
-        # Calling this should only be necessary if additional custom fields are made available
-        # after custom_field_values has already been called as that method will also build custom values
-        # (with their default values set) for all custom values for which no prior value existed.
-        def set_default_values!
```
Why can this be removed?

ulferts commented on lib/api/v3/groups/group_representer.rb:
```diff
@@ -33,6 +33,7 @@ module V3
     module Groups
       class GroupRepresenter < ::API::V3::Principals::PrincipalRepresenter
         include API::Decorators::LinkedResource
+        extend ::API::V3::Utilities::CustomFieldInjector::RepresenterClass
```
Now that groups also support custom fields in the API (such a simple change, glad that it is finally done), it would be good to briefly document that. 

On other endpoints we have something like this below the links table:

'Depending on custom fields defined for time entries, additional links might exist.'

And this below the properties list:

'Depending on custom fields defined for time entries, additional properties might exist.'

dombesz replied:
https://github.com/opf/openproject/pull/19455/commits/2f3d9440bc24d486eda04f06c2561a93d5aefed8

ulferts commented on lib/api/v3/work_packages/work_packages_api.rb:
```diff
@@ -82,6 +84,8 @@ class WorkPackagesAPI < ::API::OpenProjectAPI
                                                                parse_service: ::API::V3::WorkPackages::ParseParamsService,
                                                                params_modifier: ->(attributes) {
                                                                  attributes[:send_notifications] = notify_according_to_params
+                                                                 attributes[:validate_custom_fields] =
```
While it is true that `send_notifications` is also in the same section as the other properties, we have by now established a pattern in the API where attributes not properties of the resource are placed in the `_meta` section of the payload. 

E.g. see the payload for copying projects:

<img width="2726" height="2114" alt="Image" src="https://github.com/user-attachments/assets/0dd616e3-8722-440c-a3a1-3ccf7ac4f1d7" />

This helps differentiating between properties and parameters and I think it is beneficial to introduce this pattern here as well. Ideally `sendNotifications` would move as well but since the API is already out there, we cannot simply remove it.

ulferts commented on lib/api/v3/work_packages/work_packages_api.rb:
```diff
@@ -82,6 +84,8 @@ class WorkPackagesAPI < ::API::OpenProjectAPI
                                                                parse_service: ::API::V3::WorkPackages::ParseParamsService,
                                                                params_modifier: ->(attributes) {
                                                                  attributes[:send_notifications] = notify_according_to_params
+                                                                 attributes[:validate_custom_fields] =
```
A similar statement is missing for `POST api/v3/work_packages/:id/form` so that the client can switch validation on and off there as well.

dombesz replied:
Since moving the `validate_custom_fields` flag to the payload representer using the "_meta" tag, this is not an issue anymore.

ulferts reviewed (changes requested):
For a PR of this size and complexity I found remarkably little to complain about, @dombesz. Admittedly, I didn't go through the specs in detail but rather focused on the production code and manual testing.

I found something but that shouldn't be hard to address. On top of what is mentioned next to the code, I found the following things:

* There is an asterisks displayed after a calculated custom field marked as required for all projects in the modal that opens when clicking on the overview page to edit the fields. Since the notation changed now, so that it means to not being able to disable the cf from the project, but not to mean that a value is required for the field, I would not have expected the asterisks after the label:

<img width="956" height="186" alt="Image" src="https://github.com/user-attachments/assets/18f2b67c-906c-480c-b6cd-af8bc62681ed" />

* In two cases, the workflow is still interrupted in case of a missing or faulty cf. That happens when selecting to "Duplicate in another project" from an individual work package or from the context menu of the work package list. In this case, since the work packages will be new ones, a correct custom field is rightfully required. But the form does not provide a field for the custom fields. 
This shortcoming is out of scope for this feature. I'll open a bug for it.

dombesz committed:
Simplify the strategy apply method, and make sure it is being included as the last module.

Including the CustomActions::Actions::Strategies::CustomField as last
will make sure the custom field validation is applied correctly by using
the correct apply method.

dombesz committed:
Add groups custom field api documenation.

dombesz committed:
Move validateCustomFields to be a meta attribute in the api payload.

ulferts committed:
Merge remote-tracking branch 'origin/dev' into feature/63550-allow-editing-of-individual-work-package-project-attributes-even-if-certain-other-attributes-are-invalid-eg-required-field-empty

dombesz committed:
Revert "Simplify the strategy apply method, and make sure it is being included as the last module."

This reverts commit 4120b724ee2cbe26d2dcbc4b7311591cf71e8db9.

ulferts reviewed (approved):
With the `validateCustomFields` moved to `_meta` and available on the form endpoint as well, the main issues have been sorted out. Good work @dombesz .
